src/controlplane/client/authenticator.go

```diff
@@ -0,0 +1,133 @@
+package client
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	v1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+)
+
+// Authenticator provides an interface to generate a authorized round tripper.
+type Authenticator interface {
+	Authenticate(endpoint config.Endpoint) (http.RoundTripper, error)
+}
+
+type authenticator struct {
+	logger                  log.Logger
+	secretListerByNamespace map[string]v1.SecretNamespaceLister
+	inClusterConfig         *rest.Config
+}
+
+// NewAuthenticator returns an Authenticator that supports plain, bearer token and mTLS.
+func NewAuthenticator(logger log.Logger, secretListerByNamespace map[string]v1.SecretNamespaceLister, inClusterConfig *rest.Config) Authenticator {
+	return &authenticator{
+		logger:                  logger,
+		secretListerByNamespace: secretListerByNamespace,
+		inClusterConfig:         inClusterConfig,
+	}
+}
+
+// Authenticate retruns a round tripper according to the endpoint config.
+// For mTLS configuration it fetches the certificates from the secret.
+func (a authenticator) Authenticate(endpoint config.Endpoint) (http.RoundTripper, error) {
+	transportConfig := &transport.Config{}
+	tlsConfig := transport.TLSConfig{
```

Comment:
I would avoid a new variable, and instead I would place `transport.TLSConfig` directly inside the `transportConfig` so that in the cases we can modify the initial object and avoid `transportConfig.TLS = tlsConfig`.

I find it more readable, but maybe an other person would not